device/sycl/src/utils/barrier.hpp

```diff
@@ -8,40 +8,64 @@
 #pragma once
 
 // Project include(s).
-#include "traccc/definitions/qualifiers.hpp"
+#include "traccc/device/concepts/barrier.hpp"
 
-// SYCL includes
+// SYCL include(s).
 #include <sycl/sycl.hpp>
 
-namespace traccc::sycl {
+namespace traccc::sycl::details {
 
-struct barrier {
-    barrier(::sycl::nd_item<1> item) : m_item(item) {}
+/// A barrier type for SYCL kernels.
+template <std::size_t DIMENSIONS>
+requires(DIMENSIONS >= 1 && DIMENSIONS <= 3) struct barrier {
 
-    TRACCC_DEVICE
-    void blockBarrier() { m_item.barrier(); }
+    /// Dimensions of the barrier
+    static constexpr std::size_t dimensions = DIMENSIONS;
 
-    TRACCC_DEVICE
-    bool blockAnd(bool predicate) {
-        m_item.barrier();
+    /// Constructor
+    ///
+    /// @param item The @c ::sycl::nd_item<1> to use for the barrier.
+    ///
+    explicit barrier(const ::sycl::nd_item<dimensions>& item) : m_item(item) {}
+
+    /// @name Function(s) implementing @c traccc::device::concepts::barrier
+    /// @{
+
+    inline void blockBarrier() const {
+        ::sycl::group_barrier(m_item.get_group());
+    }
+
+    inline bool blockAnd(bool predicate) const {
+        blockBarrier();
         return ::sycl::all_of_group(m_item.get_group(), predicate);
     }
 
-    TRACCC_DEVICE
-    bool blockOr(bool predicate) {
-        m_item.barrier();
+    inline bool blockOr(bool predicate) const {
+        blockBarrier();
         return ::sycl::any_of_group(m_item.get_group(), predicate);
     }
 
-    TRACCC_DEVICE
-    unsigned int blockCount(bool predicate) {
-        m_item.barrier();
+    inline unsigned int blockCount(bool predicate) const {
+        blockBarrier();
         return ::sycl::reduce_over_group(m_item.get_group(),
                                          predicate ? 1u : 0u, ::sycl::plus<>());
     }
 
+    /// @}
+
     private:
-    ::sycl::nd_item<1> m_item;
+    /// Item object coming from the SYCL kernel
+    const ::sycl::nd_item<dimensions>& m_item;
 };
 
-}  // namespace traccc::sycl
+/// Template deduction guide for the barrier type.
+template <int N>
+barrier(::sycl::nd_item<N>) -> barrier<N>;
```

Comment:
Template deduction guides are much more fragile than I thought. :confused: When using `std::size_t` here, Clang told me that the expression "it's trying to figure out" uses `int`, so it can't use the guide with `std::size_t`.

I then tried to use the `std::integer` concept as part of the deduction guide, but that also failed. Since in that case it's the integer type that becomes the template parameter, not the integer value.

So for now I just used `int`. We may need to add other "overloads" later on as well, let's see.

Reply:
Actually this makes total sense: if you look at the template definition of `nd_item` it is indeed templated on `https://github.khronos.org/SYCL_Reference/iface/nd_item.html`.